source/game_sa/IdleCam.cpp

```diff
@@ -31,82 +34,321 @@ CIdleCam::CIdleCam() {
+    SetTarget(FindPlayerPed());
+    m_nForceAZoomOut = true;
 }
 
 // 0x517870
 void CIdleCam::ProcessTargetSelection() {
-    plugin::CallMethod<0x517870, CIdleCam*>(this);
-}
+    auto timeDelta = static_cast<float>(CTimer::GetTimeInMS()) - m_TimeLastTargetSelected;
+    if (m_ZoomState != eIdleCamZoomState::UNK_3 && m_TargetLOSCounter <= 0) {
+        timeDelta /= m_IncreaseMinimumTimeFactorForZoomedIn;
+    }
+
+    if (timeDelta > m_TimeMinimumToLookAtSomething) {
+        g_InterestingEvents.InvalidateNonVisibleEvents();
+        auto* event = g_InterestingEvents.GetInterestingEvent();
+
+        if (event) {
+            auto* eventEntity = event->entity;
+            if (m_Target != eventEntity) {
+                if (IsTargetValid(eventEntity)) {
+                    if (m_ZoomState == eIdleCamZoomState::UNK_3) {
+                        SetTarget(eventEntity);
+                    } else {
+                        m_nForceAZoomOut = true;
+                    }
+                }
+            } else if (!IsTargetValid(eventEntity)) {
+                g_InterestingEvents.InvalidateEvent(event);
+            }
+        } else if (!m_Target || !IsTargetValid(m_Target) && m_Target != FindPlayerPed()) {
+            if (m_ZoomState == eIdleCamZoomState::UNK_3) {
+                SetTargetPlayer();
+            } else {
+                m_nForceAZoomOut = true;
+            }
+        }
+    }
+
+    if (!m_Target) {
+        SetTargetPlayer();
+    }
 
-// inlined, see ProcessSlerp
-void CIdleCam::VectorToAnglesRotXRotZ(CVector* posn, float* outA1, float* outA2) {
-    assert(false);
+    if (!IsTargetValid(m_Target) && timeDelta > m_TimeMinimumToLookAtSomething) {
+        m_nForceAZoomOut = true;
+        if (m_ZoomState == eIdleCamZoomState::UNK_3 || m_TargetLOSCounter > 0) {
+            SetTargetPlayer();
+        }
+    }
+
+    if (m_TargetLOSCounter > m_TargetLOSFramestoReject) {
+        SetTargetPlayer();
+    }
 }
 
 // 0x5179E0
-void CIdleCam::ProcessSlerp(float* outA1, float* outA2) {
-    plugin::CallMethod<0x5179E0, CIdleCam*, float*, float*>(this, outA1, outA2);
+float CIdleCam::ProcessSlerp(float& outX, float& outZ) {
+    const auto beginTime = CTimer::GetTimeInMS();
+
+    CVector lookAtPos{};
+    if (m_TargetLOSCounter >= m_TargetLOSFramestoReject) {
+        lookAtPos = m_LastIdlePos;
+    } else {
+        GetLookAtPositionOnTarget(m_Target, lookAtPos);
+    }
+
+    auto [slerpAtan, slerpDistAtan]   = VectorToAnglesRotXRotZ(m_PositionToSlerpFrom - m_Cam->m_vecSource);
+    auto [lookAtAtan, lookAtDistAtan] = VectorToAnglesRotXRotZ(lookAtPos - m_Cam->m_vecSource);
+
+    const auto ClampAngle = [](float& angle, float compare) {
+        if (compare <= DegreesToRadians(180.0f)) {
+            if (compare < -DegreesToRadians(180.0f)) {
+                angle += DegreesToRadians(360.0f);
+            }
+        } else {
+            angle -= DegreesToRadians(360.0f);
+        }
```

Comment:
`fmodf`?

Reply:
i recall testing it with no success. i won't try again for now.
```suggestion
        // TODO: simplify
        if (compare <= DegreesToRadians(180.0f)) {
            if (compare < -DegreesToRadians(180.0f)) {
                angle += DegreesToRadians(360.0f);
            }
        } else {
            angle -= DegreesToRadians(360.0f);
        }
```

Reply:
possibly because `fmodf` doesn't preserve the sign the way you might expect it to, no?